highfive/newpr.py

```diff
@@ -92,7 +92,6 @@ def api_req(method, url, data=None, username=None, token=None, media_type=None):
     return { "header": header, "body": body }
 
 def post_comment(body, owner, repo, issue, user, token):
-    global post_comment_url
```

Comment:
Excuse my Python-ignorance, but why is it OK to remove this declaration? We use the variable on line 97, so it seems like it should be necessary.

Reply:
In short, you only need the `global` if you want to assign to the variable, which we aren't doing here. From [this FAQ](https://docs.python.org/2.7/faq/programming.html#what-are-the-rules-for-local-and-global-variables-in-python):
> In Python, variables that are only referenced inside a function are implicitly global. If a variable is assigned a value anywhere within the function’s body, it’s assumed to be a local unless explicitly declared as global.
>
> Though a bit surprising at first, a moment’s consideration explains this. On one hand, requiring `global` for assigned variables provides a bar against unintended side-effects. On the other hand, if `global` was required for all global references, you’d be using `global` all the time. You’d have to declare as global every reference to a built-in function or to a component of an imported module. This clutter would defeat the usefulness of the `global` declaration for identifying side-effects.

Here's an example.
```python
v = 'original'

def foo():
    # v is not global, so this creates a new, local v in foo
    v = 'foo'

def bar():
    global v
    v = 'bar'
```

```python
>>> print v
original

>>> foo()
>>> print v
original

>>> bar()
>>> print v
bar
```

Also, (not necessarily relevant to your question, but in case you're wondering as a language expert and all) it's illegal to declare something global after referencing it.
```python
>>> def baz():
...     v = 'baz1'
...     global v
...     v = 'baz2'
...
<stdin>:3: SyntaxWarning: name 'v' is assigned to before global declaration
```

Reply:
Thanks, that's a great explanation!